staging/src/k8s.io/client-go/tools/remotecommand/remotecommand.go

```diff
@@ -106,7 +113,13 @@ func NewSPDYExecutorForProtocols(transport http.RoundTripper, upgrader spdy.Upgr
 // Stream opens a protocol streamer to the server and streams until a client closes
 // the connection or the server disconnects.
 func (e *streamExecutor) Stream(options StreamOptions) error {
-	req, err := http.NewRequest(e.method, e.url.String(), nil)
+	return e.StreamWithContext(context.Background(), options)
+}
+
+// StreamWithContext opens a protocol streamer to the server and streams until a client closes
+// the connection or the server disconnects or the context is done.
+func (e *streamExecutor) StreamWithContext(ctx context.Context, options StreamOptions) error {
+	req, err := http.NewRequestWithContext(ctx, e.method, e.url.String(), nil)
 	if err != nil {
 		return fmt.Errorf("error creating request: %v", err)
 	}
```

Comment:
are there still parts inside spdy.Negotiate that can get stuck without honoring the context (specifically inside the `upgrader.NewConnection(resp)` call... will `ioutil.ReadAll(resp.Body)` unblock if the request context is closed)?

Reply:
Yes, good question. Actually I've found another case that the context isn't being honored exactly in spdy.Negotiate, but not in upgrader.NewConnection. I've dived in to the stdlib's  http client code and found that resp.Body will always be wrapped with timeout in the following code:

```go
func send(ireq *Request, rt RoundTripper, deadline time.Time) (resp *Response, didTimeout func() bool, err error) {
        ....
	if !deadline.IsZero() {
		resp.Body = &cancelTimerBody{
			stop:          stopTimer,
			rc:            resp.Body,
			reqDidTimeout: didTimeout,
		}
	}
        ...
}
```

The problem lays in the customized http.Transport returned by apimachinery/pkg/util/httpstream/spdy.NewRoundTripperWithConfig which is used in our SPDY executor. It doesn't recognize the context in request and performs lots of the new http requests & response reading without deadline. A simple way to reproduce is performing exec (timeout 1s) frequently (~ 2s) on a single pod and then shutdown it's host. With high probability you will find a stuck exec which typically timeouts after 30 seconds.

![image](https://user-images.githubusercontent.com/8898435/128803389-f2c33edc-a893-4656-9624-c32d4155315f.png)

I have a fix just for my project but isn't ready for submitting a PR to upstream. (I've changed some of the public APIs' signatures). As a reference, you can find it [here](https://github.com/arkbriar/apimachinery/compare/v0.21.2...v0.21.2-enc2).

Anyway, I think it's another issue (transport not respecting context) outside this PR (provide a way to support cancelable).

Reply:
cc. @aojea 😉

Reply:
@liggitt Hi dude, any updates?

Reply:
Sorry for the delay. I can buy that this is an improvement even if there are still parts of the operation that can't be interrupted.

Since this is a library that can be used in a lot of different ways, to be sure this doesn't introduce any regressions, asked a couple questions about behavior around panic and hanging goroutines (would like to see those behaviors exercised with unit tests, not just verified manually once)